### ISSUE TITLE
# Bug Report: cixbuild produces a degenerate/collapsed model when quantizing an OSNet re-ID (person re-identification) checkpoint

### ISSUE THREAD

**Issue by alhaddad-m** [KaiyangZhou/deep-person-reid#600]
## Summary

Quantizing an OSNet (`osnet_x1_0` / `osnet_ain_x1_0`, from the [torchreid](https://github.com/KaiyangZhou/deep-person-reid) library) ONNX graph with `cixbuild` produces a `.cix` model whose output is **numerically collapsed** (near-constant regardless of input) whenever the underlying PyTorch weights come from a **person re-ID (triplet/ID-loss) trained checkpoint**. The exact same export/quantization pipeline, config, and target, applied to the **same architecture with ImageNet-classification-trained weights**, produces a correctly-functioning model. We have reproduced this failure with **three independent checkpoints** and isolated it to the `cixbuild` step specifically — everything upstream (PyTorch inference, ONNX export, `onnxsim` simplification) has been verified numerically correct.

## Environment

- Toolchain: CIX NOE SDK, `cixbuild` — build reports `Build with version 6.1.3407`, `Compass-Optimizer, version: 1.3.3407`
- Target: `X2_1204MP3` (CIX P1 NPU, used on Orion O6/O6N)
- Runtime on-device: `libnoe` / `NOE_Engine` Python API (`noe_load_graph`, `noe_job_infer_sync`, etc.)
- Model source: `torchreid` (`pip install torchreid`), PyTorch, `torch.onnx.export` (opset 12), `onnxsim`

## Pipeline (identical for working and failing cases)

1. `torchreid.models.build_model(name=<model_name>, num_classes=1000, pretrained=False)`
2. `torchreid.utils.load_pretrained_weights(model, <checkpoint_path>)`
3. `model.eval()`
4. `torch.onnx.export(model, dummy_input, ..., input_names=["input"], output_names=["embedding"], opset_version=12, do_constant_folding=True, dynamic_axes=None)` — input shape `[1, 3, 256, 128]`
5. `onnxsim.simplify()` on the exported graph
6. `cixbuild` with the following `[Optimizer]` config:
   ```
   quantize_method_for_weight = per_channel_symmetric_restricted_range
   quantize_method_for_activation = per_tensor_asymmetric
   weight_bits = 16
   activation_bits = 16
   bias_bits = 32
   calibration_batch_size = 1
   ```
   Calibration data: ~100–300 representative person-crop images (JPEG), preprocessed with standard ImageNet mean/std normalization at 256×128, saved as a `.npy` array (`numpydataset`).
7. Output loaded on-device via `NOE_Engine` (`libnoe`) and run through `noe_job_infer_sync`.

## Working baseline

`osnet_x1_0` built via `torchreid.models.build_model(..., pretrained=True)` (torchreid's built-in **ImageNet classification** checkpoint download, not a re-ID checkpoint), run through the exact same steps 4–7 above. The resulting `.cix` file works correctly: output varies sensibly per input, and person re-identification similarity scores are consistent with expectations (same person across two photos: cosine ≈ 0.87–0.92; different people: cosine ≈ 0.63–0.87, clearly separated).

## Failing cases (3 independently reproduced, identical symptom)

| # | Architecture | Checkpoint | Source |
|---|---|---|---|
| 1 | `osnet_x1_0` | `osnet_x1_0_msmt17_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip.pth` | torchreid Model Zoo, MSMT17 (single-domain re-ID) |
| 2 | `osnet_ain_x1_0` | `osnet_ain_ms_d_c.pth.tar` | torchreid Model Zoo, multi-source domain-generalization re-ID (MSMT17+DukeMTMC+CUHK03 → Market1501) |
| 3 | `osnet_x1_0` | `osnet_x1_0_market_256x128_amsgrad_ep150_stp60_lr0.0015_b64_fb10_softmax_labelsmooth_flip.pth` | torchreid Model Zoo, Market-1501 (single-domain re-ID) |

All three, after `cixbuild`, produce a `.cix` model with **the same collapse symptom**:

- Feeding 5 unrelated random-noise images (uint8, `[300,150,3]`, uncorrelated `numpy.random` draws) through the deployed model yields output embedding vectors with **pairwise cosine similarity of 0.996–0.999** (i.e. essentially the same vector regardless of input). The first several output components are near-identical across all 5 inputs (e.g. `[0.0, 0.0821, 0.1179, 0.0, 0.0]`, `[0.0, 0.0857, 0.1172, 0.0, 0.0]`, ...).
- Feeding real photos of 4 and 3 clearly different people (two separate photographs) yields cross-photo cosine similarities of **0.87–0.99 for every pair**, including pairs that are confirmed different individuals — i.e. no usable identity signal survives quantization+deployment.

For comparison, the working ImageNet-weights build, tested identically, gives 0.63–0.87 for different people and 0.87–0.92 for the same person — a real, usable margin.

## Isolation work performed (rules out our export code / weights)

1. **Raw PyTorch model (before any export)**: verified healthy for all 3 checkpoints. `model(input).shape == [N, 512]`, reasonable output statistics (e.g. for the Market-1501 checkpoint: mean≈0.56, std≈0.55, min=0, max≈2.9 — normal post-ReLU feature magnitudes), and correct nearest-neighbor person re-identification on a controlled test (crops of the same 3 people across two photographs): Market-1501 checkpoint gave 3/3 correct top-1 matches (cosine 0.786–0.865 for true matches vs 0.532–0.828 for non-matches); `osnet_ain` gave 3/3 correct; MSMT17 gave 2/3 correct — none showed anything resembling collapse.
2. **ONNX export + `onnxsim`**: the resulting `-sim.onnx` file, run through `onnxruntime` (`CPUExecutionProvider`, full FP32, no quantization) on the same test crops, reproduces the **raw PyTorch cosine-similarity matrix exactly to 3 decimal places** (Market-1501 case verified). This rules out any issue in `torch.onnx.export` or `onnxsim`.
3. **Calibration statistics** (`*_statistic_info.npy`, produced with `save_statistic_info = True`): compared per-tensor `extrema_min/max`, `running_min/max/mean/std/mad` (and per-channel `_key_axis` variants) between the working ImageNet build and a failing build (same `osnet_x1_0` architecture, only weights differ — MSMT17). Found **no NaN/Inf anywhere** in either build, and no per-tensor value blowups beyond what looks like ordinary variation between two differently-trained checkpoints of the same architecture (e.g. differing channel-attention gate activation ranges) — nothing that looks like numeric overflow at the calibration level.
4. **`activation_bits` cannot be raised above 16** to further isolate activation-quantization sensitivity — the Optimizer's config validator rejects it: `Required the integer 'activation_bits' field and must be in [8, 9, 10, 11, 12, 13, 14, 15, 16], now is <class 'int'> type, 32, default value=8.`

## A specific anomaly worth investigating

On one of the failing builds (`osnet_ain_x1_0` + multi-source checkpoint), the `[OPT]` optimize stage logged:

```
.../AIPUBuilder/Optimizer/framework/pycore/pyir.py:923: RuntimeWarning: invalid value encountered in cast
  ct_value = ct_value.astype(dtype2nptype(ct.dtype))
...
[I] [OPT] [10:05:52]: [Done]cost time: 1602s, and [qinfos(scale, zp, dtype)]: out: [[19691.595703125, 0, UINT16]] in: [[12491.517578125, 0, INT16]] [output tensors cosine]: [0.394930563107717][output tensors MSE]: [2.306063652038574]
```

The Optimizer's own internal float-vs-quantized fidelity check reports a cosine similarity of only **0.395** between the quantized and reference-float representations of (what appears to be) the final output tensor. This is consistent with, and is likely the direct cause of, the on-device collapse we observe. We were not able to determine which layer produces the `invalid value encountered in cast` warning, or whether this cosine=0.395 check corresponds specifically to the network's final output or an intermediate tensor.

## Architecture note

Both `osnet_x1_0` and `osnet_ain_x1_0` contain a channel-attention "Unified Aggregation Gate" block per residual stage: `GlobalAveragePool → Conv (fc1) → ReLU → Conv (fc2) → Sigmoid`. This is the only place in the network where a `Sigmoid` activation appears, and it is present in both the working and failing builds — we flag it as one plausible (unconfirmed) location for quantization sensitivity, but have not been able to confirm or rule it out with the tools available to us.

## Questions for CIX engineering

1. Is this a known limitation/issue with `cixbuild`/Compass-Optimizer when quantizing networks whose weight/activation statistics come from a metric-learning (triplet/ID-loss) trained checkpoint, as opposed to a standard classification-trained checkpoint of the same architecture?
2. Given `activation_bits` is capped at 16 via config, is there a way to run a genuinely float/unquantized activation pass (or a much higher-precision debug pass) to confirm whether activation quantization specifically is the cause?
3. Is per-channel activation quantization (as opposed to `per_tensor_asymmetric`) supported, and would it be expected to help here?
4. Is a different activation calibration strategy than `mean` available (e.g. percentile- or KL-divergence-based), and could that plausibly resolve this?
5. Could CIX engineering run the attached ONNX file through an internal reference quantizer/simulator to compare against what we're seeing, and/or identify which specific node the `invalid value encountered in cast` warning originates from?

## Attachments available on request

- `osnet_x1_0_market1501-sim.onnx` (post-`onnxsim`, verified numerically correct via `onnxruntime`)
- `osnet_build.cfg` used for the failing build
- `osnet_x1_0_statistic_info.npy` for both a working (ImageNet) and failing (MSMT17) build of the same architecture
- The ~100–300 calibration crop images used
- The Python reproduction scripts (export → onnxsim → cixbuild, and the on-device noise-collapse sanity check)
